### ISSUE TITLE
Continue fetching key packages directly?

### ISSUE THREAD

**Issue by turt2live** [bifurcation/ietf-mimi-protocol#33]
> **TODO**: For now, we assume that KeyPackages are fetched directly, i.e. not in the
> context of a room and via a Hub. This might change in the future. If it does
> change, this event needs an additional authentication mechanism.

**Comment by turt2live** [comment]
In previous discussions, the working group indicated that it's preferred to route everything through the hub: does this still hold true? 

Proposal from my side would be to route traffic through the hub, though authentication of the proxied material may require direct pokes. This would likely apply for #24 as well (with authentication being a hash in the message, to prevent modified downloads)

**Comment by beltram** [comment]
tl;dr: fetch KeyPackages from the Hub.

Consider the case of expirable credentials (X509, VCs): they have to be rotated at some point.

If a KeyPackage is not fetched from the Hub consistently then a client might fetch a KeyPackage which corresponding private key has been deleted by the emitter. This would lead to the creation of a Welcome whose receiver would be unable to decrypt, becoming a ghost client.

**Comment by kkohbrok** [comment]
I don't see where the routing (i.e. via the Hub or directly) makes a difference in the context of KeyPackage or Credential expiration. In both cases the KeyPackage from Bob's server will reach Alice at almost the same time.

From what I understood, the biggest concern is that direct fetching "violates" star-shaped topology and can thus lead to situations, where Alice wants to fetch a KeyPackage from Bob, but her provider doesn't allow direct contact. Alice would thus have to fetch through a Hub. However, if Alice and Bob can only connect through that Hub, the Alice probably shouldn't be able to obtain KeyPackages from Bob in the first place.

For authentication key material, I suggest (for now) we assume that the credential in each KeyPackage contains a full chain that everyone can verify against their local (web) root of trust.

**Comment by beltram** [comment]
Here is the scenario I have in mind. If we want to strictly prevent ghost clients we have to fetch KeyPackages from the Hub. We could relax that rule for Basic KeyPackages or in the presence of a `PreventGhostClients=false` RoomPolicy.

```mermaid
sequenceDiagram
  participant B as Bob
  participant HP as Hub Provider
  participant GP as Guest Provider
  participant A as Alice
  B ->>+ HP: Replace its KeyPackages
  HP ->>- B: Ok KeyPackages replaced
  B ->> B: Deletes asociated private keys
  A ->>+ GP: Fetch Bob (old) KeyPackage
  GP ->>- A: Here is 1 KeyPackage from Bob
  HP ->>+ GP: Bob's KeyPackage have changed
  A ->> A: Create Add Commit + Welcome
  A ->> GP: Send Welcome
  GP ->> HP: Fanout Welcome
  HP ->> B: Receive Welcome
  B ->> B: Decryption error
```

**Comment by kkohbrok** [comment]
What I don't understand is why the guest provider would buffer Bob's KeyPackages instead of just forwarding Alice's Query to Bob's server (which in this case is the Hub)? KeyPackages are (for the most part) single-use. Or are you talking about last-resort KeyPackages?

**Comment by beltram** [comment]
You're right it does not make sense for Alice's Provider to buffer other Provider's KeyPackage. Let me redact my example. Here if the Hub is not responsible for the KeyPackages of the room, then it would validate the Add Commit of an already deleted KeyPackage. I concede this example is an edge-case but it could still occur.

```mermaid
sequenceDiagram
  participant B as Bob
  participant GPB as Guest Provider Bob
  participant HP as Hub Provider
  participant GPA as Guest Provider Alice
  participant A as Alice
  A ->>+ GPB: Fetch Bob KeyPackage
  GPB ->>- A: Bob KeyPackage
  A ->> A: Create Add Commit + Welcome
  B ->>+ GPB: Replace its KeyPackages
  A ->> HP: Send Add Commit + Welcome
  GPB ->>- B: Ok KeyPackages replaced
  B ->> B: Deletes asociated private keys
  HP ->> GPB: Fanout Welcome
  GPB ->> B: Receive Welcome
  B ->> B: Decryption error
```

**Comment by kkohbrok** [comment]
Ah, I see now what you mean, thanks! A few comments:

If we follow what I think you suggest, then Bob would have to upload KeyPackages to every Hub he is involved in and, more generally, Hubs would (have to) manage key material for every client that is in one of their groups. This introduces quite a bit of complexity as Bob now has to track where he uploaded which KeyPackages.

Even if we implement that solution, we would still have the same problem in cases where Bob doesn't have any KeyPackages on Alice's Hub (because he is not a participant of any groups there). Then Alice has to fetch KeyPackages from Bob's server (through her Hub) and Bob can replace that KeyPackage while Alice is creating the Add, leading to the same situation.

I don't think we can completely solve this issue. However, Bob can recover from this situation by doing an external commit. He may have missed a few messages, but he can recover. Also, Bob can mitigate this issue by keeping the key material around for a short while after replacing KeyPackages on the server. After all, Alice is not going to wait a very long time between fetching a KeyPackage and performing the Add operation.

To summarize, I don't think the problem is that bad, because there is a mitigation and a recovery mechanism. Instead of taking the complexity hit of keeping KeyPackages on every Hub, we should instead focus on a simpler solution (fetching via the Hub, or fetching directly from Bob's server).

**Comment by dkasak** [comment]
It seems to me that the problem presented here would largely be avoided if instead of key package replacement we were talking about key package *replenishment*.

Replacement implies we're dropping already published but supposedly unused key packages, which then turn out to have actually been used, triggering the issue. On the other hand, replenishment (up to a given max total) never drops already published key packages before they're used, only adds new ones. This then mostly avoids the problem of dropping key packages prematurely.

Under this view, the key package fetching operation is to be considered to be consuming the key package, thereby reducing the number of available key packages on the guest provider, instructing the user (Bob in the diagram above) to replenish a larger number of key packages in order to reach the max total.

**Comment by kkohbrok** [comment]
While it would certainly make things simpler, I don't think we can assume that clients will never need to remove/revoke KeyPackages. For example, if a client rotates its credential for some reason, it won't want to wait until all previously published KeyPackages have expired.

**Comment by dkasak** [comment]
You're right, of course. A client could also simply lose its private key material for the published key packages due to an error or failure. This makes it obvious removal needs to be supported.

However, I maintain that the default case should be to non-destructively replenish rather than replace, and for the fetching operation to be considered to be consuming (reducing the available total), which in practice makes this problem *much* less likely to occur.

**Comment by tgeoghegan** [comment]
Key packages are now delivered via the hub, so this is resolved.
